### PR TITLE
Add defaultFontSizes option to theme.json

### LINKED ISSUE #52200: Global styles: Default font sizes are added
### Description

I created a blank theme and added custom font sizes.
This works intended in the editor, but in the Global styles, default font sizes are added in the drop down (S, M, L, XL)

### Step-by-step reproduction instructions

1. Copy the nearly blank theme.js from below
2. Go into Global Style → Typography → Text
3. See in the drop down the additional "Small, Medium, Large, Extra Large" font sizes, that shouldn't be there.

<details><summary>theme.js</summary>
<p>

```javascript
{
  "settings": {
    "appearanceTools": true,
    "color": {
      "background": true,
      "custom": true,
      "customDuotone": true,
      "customGradient": true,
      "defaultDuotone": false,
      "defaultGradients": false,
      "defaultPalette": false,
      "duotone": [],
      "gradients": [],
      "palette": [],
      "text": true
    },
    "layout": {
      "contentSize": "620px",
      "wideSize": "1000px"
    },
    "shadow": {
      "defaultPresets": true,
      "presets": []
    },
    "spacing": {
      "customSpacingSize": true,
      "spacingScale": {
        "increment": 1.5,
        "mediumStep": 1.5,
        "operator": "*",
        "steps": 7,
        "unit": "rem"
      },
      "spacingSizes": [],
      "units": ["%", "px", "em", "rem", "vh", "vw"]
    },
    "typography": {
      "customFontSize": true,
      "dropCap": true,
      "fluid": true,
      "fontFamilies": [
        {
          "fontFamily": "-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
          "name": "System Font",
          "slug": "system-font"
        },
        {
          "fontFace": [
            {
              "fontFamily": "Barlow",
              "fontStyle": "italic",
              "fontWeight": "700",
              "src": ["file:./assets/fonts/barlow_italic_700.ttf"]
            },
            {
              "fontFamily": "Barlow",
              "fontStyle": "normal",
              "fontWeight": "700",
              "src": ["file:./assets/fonts/barlow_normal_700.ttf"]
            },
            {
              "fontFamily": "Barlow",
              "fontStyle": "italic",
              "fontWeight": "600",
              "src": ["file:./assets/fonts/barlow_italic_600.ttf"]
            },
            {
              "fontFamily": "Barlow",
              "fontStyle": "normal",
              "fontWeight": "600",
              "src": ["file:./assets/fonts/barlow_normal_600.ttf"]
            },
            {
              "fontFamily": "Barlow",
              "fontStyle": "italic",
              "fontWeight": "400",
              "src": ["file:./assets/fonts/barlow_italic_400.ttf"]
            },
            {
              "fontFamily": "Barlow",
              "fontStyle": "normal",
              "fontWeight": "400",
              "src": ["file:./assets/fonts/barlow_normal_400.ttf"]
            }
          ],
          "fontFamily": "Barlow",
          "slug": "barlow"
        }
      ],
      "fontSizes": [
        {
          "name": "Display xl",
          "slug": "Display-xl",
          "size": "96px",
          "fluid": {
            "max": "96px",
            "min": "52px"
          }
        },
        {
          "name": "Display lg",
          "slug": "Display-lg",
          "size": "72px",
          "fluid": {
            "max": "72px",
            "min": "46px"
          }
        },
        {
          "name": "Display md",
          "slug": "Display-md",
          "size": "60px",
          "fluid": {
            "max": "60px",
            "min": "42px"
          }
        },
        {
          "name": "Display sm",
          "slug": "Display-sm",
          "size": "52px",
          "fluid": {
            "max": "52px",
            "min": "40px"
          }
        },
        {
          "name": "Heading 1",
          "slug": "Heading-1",
          "size": "48px",
          "fluid": {
            "max": "48px",
            "min": "36px"
          }
        },
        {
          "name": "Heading 2",
          "slug": "Heading-2",
          "size": "36px",
          "fluid": {
            "max": "36px",
            "min": "30px"
          }
        },
        {
          "name": "Heading 3",
          "slug": "Heading-3",
          "size": "28px",
          "fluid": {
            "max": "28px",
            "min": "26px"
          }
        },
        {
          "name": "Heading 4",
          "slug": "Heading-4",
          "size": "24px",
          "fluid": {
            "max": "24px",
            "min": "22px"
          }
        },
        {
          "name": "Heading 5",
          "slug": "Heading-5",
          "size": "20px",
          "fluid": {
            "max": "20px",
            "min": "19px"
          }
        },
        {
          "name": "Heading 6",
          "slug": "Heading-6",
          "size": "18px",
          "fluid": {
            "max": "18px",
            "min": "17px"
          }
        },
        {
          "name": "Paragraph lg",
          "slug": "paragraph-lg",
          "size": "18px",
          "fluid": {
            "max": "18px",
            "min": "17px"
          }
        },
        {
          "name": "Paragraph base",
          "slug": "paragraph-base",
          "size": "16px",
          "fluid": {
            "max": "16px",
            "min": "16px"
          }
        },
        {
          "name": "Paragraph sm",
          "slug": "paragraph-sm",
          "size": "14px",
          "fluid": {
            "max": "14px",
            "min": "14px"
          }
        },
        {
          "name": "Info md",
          "slug": "Info-md",
          "size": "12px",
          "fluid": {
            "max": "12px",
            "min": "12px"
          }
        },
        {
          "name": "Info sm",
          "slug": "Info-sm",
          "size": "10px",
          "fluid": {
            "max": "10px",
            "min": "10px"
          }
        }
      ],
      "fontStyle": true,
      "fontWeight": true,
      "letterSpacing": true,
      "textDecoration": true,
      "textTransform": true
    },
    "useRootPaddingAwareAlignments": true
  },
  "styles": {
    "spacing": {
      "padding": {
        "bottom": "0px",
        "left": "16px",
        "right": "16px",
        "top": "0px"
      }
    }
  },
  "templateParts": [
    {
      "area": "header",
      "name": "header"
    },
    {
      "area": "footer",
      "name": "footer"
    }
  ],
  "version": 2,
  "$schema": "https://schemas.wp.org/trunk/theme.json"
}
``` 

</p>
</details> 

### Screenshots, screen recording, code snippet

| Block styles (as it should be) | Global styles (with additonal font sizes) |
|--------|--------|
| image | image |

### Environment info

- WP 6.2.2
- GB 16.1.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ajlende**
## What?

Adds a `defaultFontSizes` option to disable showing the default font sizes.

When `defaultFontSizes === false` the options are removed from the UI.

By default `defaultFontSizes` is `true`.

## Why?

Resolves #52200

## How?

- Added option in the places it was needed
- Updated the base `TypographyPanel` to remove defaults when the option was set to false

## Testing Instructions

Test the new `defaultFontSizes`.

1. Add `settings.typography.defaultFontSizes = false` to your theme.json file.
2. See that the default options do not show.
3. Remove the setting.
4. See that the default options do show again.

Test that duplicate font sizes are removed.

1. Update `settings.typography.fontSizes` to include duplicate slugs.
2. Ensure that only the first of the duplicates is included.

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 

Tested with the theme.json from the example in #52200 and `defaultFontSizes` set to `false`.

| Before | After |
| ------ | ----- |
| image | image |

### CODE REVIEW

**Review by scruffian — COMMENTED** [review]
When I add a duplicate font size, the first one shows in the list but the second one is actually what gets applied.

**Comment by ajlende** [comment]
> When I add a duplicate font size, the first one shows in the list but the second one is actually what gets applied.

This looks like it exists on trunk too. Generating the global stylesheet doesn't filter out duplicates, so the later CSS overrides the earlier.

```json
{
	"name": "Display xl2",
	"slug": "Display-xl",
	"size": "296px"
},
{
	"name": "Display xl",
	"slug": "Display-xl",
	"size": "96px"
},
```

duplicates overriding one another

There's a second bug on trunk. You should be able to override default values because `prevent_override` is `false` in `PRESETS_METADATA`, but for some reason that doesn't work. The core defaults are shown and used instead.

```json
{
	"name": "Override Small",
	"slug": "small",
	"size": "26px"
},
{
	"name": "Override Medium",
	"slug": "medium",
	"size": "40px"
},
{
	"name": "Override Large",
	"slug": "large",
	"size": "72px"
},
```

It's an issue with global styles, but fixing global styles seems like a separate task, so I recreated the second bug in this PR (and fixed the first bug where overriding yourself shows only the value that will be applied in CSS). At least that way, the experience isn't any worse than it was before. The thing you select should always match the thing in the CSS even if you still can't override the defaults. The problem is that I had to write some hacky code to match the UI with the buggy global styles.

This needs more testing, and maybe some cleanup. But I'm out for a few days, so if someone wants to pick it up and bring it to the finish line while I'm away, you're more than welcome to do so. Getting the global styles `prevent_override` issue in a separate PR could also unblock this.

**Comment by scruffian** [comment]
Here's what I found when investigating the duplicate font sizes: https://github.com/WordPress/gutenberg/pull/57692

File: packages/block-editor/src/components/global-styles/typography-panel.js
Comment (scruffian) [inline packages/block-editor/src/components/global-styles/typography-panel.js]: I'm now not convinced that's what `prevent_override` is meant to do. I think the purpose of that is to allow the different settings to be split out onto different levels within the array created by theme.json (i.e `default`, `theme` and `custom`). This prevents the theme from overriding the defaults.

The question of what happens to duplicates not dealt with by Global Styles I think, but the behaviour seems consistent - whatever the preset is, is kept and cannot be overridden by the theme. I think we should keep this behaviour, which means I think we can merge this PR, we just need to update this comment.

**Comment by hanneslsm** [comment]
> This looks like it exists on trunk too. Generating the global stylesheet doesn't filter out duplicates, so the later CSS overrides the earlier.

This seems to be the same (?) bug as described in https://github.com/WordPress/gutenberg/issues/53558. Linking for visibility.

**Comment by ajlende** [comment]
Thanks for the link @hanneslsm–your issue is indeed very similar!

It seems the issue you're describing is related to duplicate `"size"` properties, whereas this issue is related to duplicate `"slug"` properties. I opened #57733 because it seemed different enough from a technical implementation point of view, but they might very well be able to be fixed at the same time.

File: packages/block-editor/src/components/global-styles/typography-panel.js
Comment (ajlende) [reply packages/block-editor/src/components/global-styles/typography-panel.js]: My comment was just poorly worded. I opened a separate issue for a follow-up (#57733) and updated the comment to be a bit more clear that the hack is to update the UI to match the buggy stylesheet.

**Review by scruffian — APPROVED** [review]
This solves the bug without making the issue with duplicate values worse, so I think we should merge it.

**Comment by getdave** [comment]
Thank you for working on this.

Once the code is merged into WP Core please could you 

- check off the line items in the [PHP Sync Tracking Issue](https://github.com/WordPress/gutenberg/issues/57959) 
- remove the `Needs PHP backport` label
- add the `Backported to WP Core` label

Thank you in advance 🙇

**Comment by getdave** [comment]
Trac - https://core.trac.wordpress.org/ticket/60284

PR - https://github.com/WordPress/wordpress-develop/pull/5896

**Comment by getdave** [comment]
I have been advised away from Github, that merging the backport is delayed here due to some concerns regarding compatibility with existing Themes. @ajlende will elaborate when he has more details.

**Comment by ajlende** [comment]
This PR changed the default value for `prevent_override` to be the value of `defaultFontSizes` which is `true` by default. Previously `prevent_override` was `false`, so now any theme defined font sizes that use the same slugs as the default font sizes (`small`, `medium`, `large`, etc.) are no longer overridden by their theme counterparts. I didn't think this was an issue because of the bug described in https://github.com/WordPress/gutenberg/pull/56661#issuecomment-1846271933. However, apparently in some cases the ordering was correct, so this change broke those sites.

The value for `defaultFontSizes` can't be set to `false` because it is also in control of the UI, and themes that don't define custom font sizes would no longer show the defaults. So I'm planning on adding a third value `"legacy"` or something that global styles can interpret as `prevent_override` being `false`, but the UI can interpret as `defaultFontSizes` being `true` so that they are all shown.

So both the ordering bug and the legacy behavior both need to be added to fix this issue which is why there are delays.

**Comment by getdave** [comment]
Thank you for the update 👍  Let either of the Editor Tech Leads know if you need any support to land the fixes.

**Comment by youknowriad** [comment]
Removing the labels as this has been reverted.

**Pull request merged**
